Improved `mig/shared/fileio.py` unit test coverage

jonasbardino opened the pull request:
Adds basic unit tests for additional `mig/shared/fileio.py` functions like:
 * `read_file`, `read_file_lines`, `read_head_lines`, `read_tail_lines`,  `write_file_lines`
 * `make_symlink`, `delete_symlink`
 * `touch`, `delete_file`, `get_file_size`
 * `makedirs_rec`, `remove_dir`, `remove_rec`,  `check_empty_dir`
 * `move_file`, `move_rec`, `copy_file`, `copy_rec`
 * `check_read_access`, `check_write_access`, `check_readable`, `check_writable`, `check_readonly`, `check_readwritable`
 reusing the existing structure.

Includes a **TODO** to fix `get_file_size` to actually return result value on exception as it currently doesn't and therefore effectively returns `None` e.g. for missing paths.

Includes a **TODO** to fix a (possible py3 regression) bug in `touch` to make it work again for existing files and directories.

Includes tests to cover the corner-case bugs of the check access helpers registered in issue #378 when run outside the docker containers e.g. with `make unittest` rather than `make test`.

jonasbardino committed:
Adds basic unit tests for:
 1. `read_file`
 2. `read_file_lines`
 3. `get_file_size`
 4. `delete_file`
 5. `read_head_lines`
 6. `read_tail_lines`
 7. `make_symlink`
reusing the existing structure.

Includes a TODO to fix get_file_size to actually return result value on
exception as it currently doesn't and therefore effectively returns None.

jonasbardino committed:
Cover fileio.write_file_lines as well reusing existing structure.

jonasbardino committed:
Extend tests to cover delete_symlink reusing make_symlink tests.

jonasbardino committed:
Adds comprehensive tests for:
1. `touch`
2. `remove_dir`
3. `remove_rec`
4. `move_file` and `move_rec`
5. `copy_file` and `copy_rec`
6. `check_empty_dir`
7. `makedirs_rec`

Discovered a corner-case bug in `touch` which renders it broken for existing
files and directories. So the corresponding tests are disabled and with TODO
markers to fix the bug and enable them.

jonasbardino committed:
Cover the various check access helpers in the fileio unit tests.

Please note that we effectively run unit tests as uid 0 when using the
containerized setup, and that may yield rather surprising results when testing
file access, because super-users have access to everything despite what the
permissions indicate.

The added tests take uid 0 into account and will succeed, but we should really
eliminate these privileges when unit testing as they can easily cause a lot of
head-scratching.

The tests also uncovered a couple of additional corner case bugs in the access
checks, which don't show up with the uid 0 case but need to be addressed for
the general benefit. They have been registered in issue #378 for further work.

albu-diku commented on tests/test_mig_shared_fileio.py:
```diff
@@ -246,5 +274,936 @@ def test_store_unicode_in_binary_mode(self):
+        # ├── file1.txt
+        # └── subdir/
+        #     └── file2.txt
+        os.makedirs(os.path.join(self.tmp_src, 'subdir'))
```
Nice detail.

albu-diku reviewed (changes requested):
The tests look ok but they all define a `setUp` function containing a super call.

Our tests provide a hook named `before_each()` for this purpose that is called at the appropriate time (i.e. after configuration is made available) and doe snot require super lines. This needs to be corrected across all cases.

jonasbardino commented on tests/test_mig_shared_fileio.py:
```diff
@@ -246,5 +274,936 @@ def test_store_unicode_in_binary_mode(self):
+        self.assertTrue(fileio.check_read_access(self.tmp_dir,
+                                                 parent_dir=True))
+        # Super-user has access to read and write all files!
+        if os.getuid() == 0:
```
It's not great that unit tests run as `root` unlike our deployment environments with an unprivileged user, so that would of course be good to get fixed. I don't know how involved that will be, however.

jonasbardino replied:
I wrote some more about it in #390 comments and marked it as follow-up there. Feel free to pursue the issue or catch me about coordinating the efforts on that front.

jonasbardino replied:
Should be sorted with #402 merged.

jonasbardino commented:
> The tests look ok but they all define a `setUp` function containing a super call.
> 
> Our tests provide a hook named `before_each()` for this purpose that is called at the appropriate time (i.e. after configuration is made available) and doe snot require super lines. This needs to be corrected across all cases.

This is what I tried to say when mentioning the apparent lack of one or more "golden standard" unit tests to use as template. This one clearly reused heavily form one of the existing test suites where `setUp` was the way it was done. I don't mind switching to `before_each` if that is indeed the way forward.

albu-diku commented:
> > The tests look ok but they all define a `setUp` function containing a super call.
> > Our tests provide a hook named `before_each()` for this purpose that is called at the appropriate time (i.e. after configuration is made available) and doe snot require super lines. This needs to be corrected across all cases.
> 
> This is what I tried to say when mentioning the apparent lack of one or more "golden standard" unit tests to use as template. This one clearly reused heavily form one of the existing test suites where `setUp` was the way it was done. I don't mind switching to `before_each` if that is indeed the way forward.

I'm sorry that this is the cause of so much frustration. For the record, before_each is not a new convention - there are 12 tests in tree that make use of it on the next branch. There is only one test which, to my surprise, has this alternative and that is test_mig_shared_fileio.py. Not sure when that crept in.

A document that begins to address the "golden standard" has been worked on and a first iteration is available. That fact has not been announced given the majority of the team has been holidaying and return tomorrow (Monday 5th January).

jonasbardino commented:
> > > The tests look ok but they all define a `setUp` function containing a super call.
> > > Our tests provide a hook named `before_each()` for this purpose that is called at the appropriate time (i.e. after configuration is made available) and doe snot require super lines. This needs to be corrected across all cases.
> > 
> > 
> > This is what I tried to say when mentioning the apparent lack of one or more "golden standard" unit tests to use as template. This one clearly reused heavily form one of the existing test suites where `setUp` was the way it was done. I don't mind switching to `before_each` if that is indeed the way forward.
> 
> I'm sorry that this is the cause of so much frustration. For the record, before_each is not a new convention - there are 12 tests in tree that make use of it on the next branch. There is only one test which, to my surprise, has this alternative and that is test_mig_shared_fileio.py. Not sure when that crept in.
> 
> A document that begins to address the "golden standard" has been worked on and a first iteration is available. That fact has not been announced given the majority of the team has been holidaying and return tomorrow (Monday 5th January).

No frustration on my side :-) ... just trying to trace and explain where this particular inconsistency crept in - and how any other such remaining "intermediate iteration" leftovers can easily multiply due to the way we tend to reuse heavily from existing tests in our unit test development workflow.
Thanks for clarifying that `before_each` is now indeed the standard and for the effort on documenting how to do tests right (TM). I'll replace `setUp` here, keep an eye on that doc and make sure to ask next time I'm unsure.

jonasbardino committed:
Rework tests to address PR review feedback (thanks @albu-diku):
 * Switch from old `setUp` to `before_each` init structure using 'testconfig'
 * Replace direct `os.makedirs` calls with the `ensure_dirs_exist` helper
 * Converge on string constants for almost everything and greater reuse
 * Strip dir-prefix part of path constants and adjust asserts accordingly

Please note that make unittest still fails the three tests of #378 yet to be
fixed by #379.

jonasbardino commented:
> The tests look ok but they all define a `setUp` function containing a super call.
> 
> Our tests provide a hook named `before_each()` for this purpose that is called at the appropriate time (i.e. after configuration is made available) and doe snot require super lines. This needs to be corrected across all cases.

I've updated to consistently use `before_each` in the latest version. Please have a look and either approve or let me know if there's anything else missing.

albu-diku reviewed (changes requested):
The tests look good - with the `uid == 0` conditionals removed throughout this can be approved and merged,

jonasbardino commented:
> The tests look good - with the `uid == 0` conditionals removed throughout this can be approved and merged,

Why would you want the inline `uid == 0` conditionals removed? They reflect the expected and correct behaviour for the tested code in that case. Namely, if somebody e.g. manually runs the unit tests as a privileged user.

I agree that we do **not** want to end up in that case running from our own `make` targets or unprivileged env runs, but I think that's a different matter.  Similarly we should gradually phase out any remaining skip-decorators blindly using that condition now that we carefully handle privileges in `make test`.

If we want to actively refuse privileged use I think that would belong e.g. as an exception in the tested function and with tests adjusted accordingly.

albu-diku commented:
> > The tests look good - with the `uid == 0` conditionals removed throughout this can be approved and merged,
> 
> Why would you want the inline `uid == 0` conditionals removed? They reflect the expected and correct behaviour for the tested code in that case. Namely, if somebody e.g. manually runs the unit tests as a privileged user.
> 
> I agree that we do **not** want to end up in that case running from our own `make` targets or unprivileged env runs, but I think that's a different matter. Similarly we should gradually phase out any remaining skip-decorators blindly using that condition now that we carefully handle privileges in `make test`.
> 
> If we want to actively refuse privileged use I think that would belong e.g. as an exception in the tested function and with tests adjusted accordingly.

Hm, ok. I had thought those checks were temporary to work around accidentally elevated privilege, my mistake if not.

However, I don't think having conditions like around the checked assertions is something we should really encourage. If there are behaviour differences for root users, and it's important to have those "documented", then I really think those ought to be separate tests asserting that circumsance/outcome. One could then much more cleanly make then conditional by using the standard skipping decorator, where you skip if uid ==0 in once case and uid !=0 in the other.

jonasbardino commented:
> > > The tests look good - with the `uid == 0` conditionals removed throughout this can be approved and merged,
> > 
> > 
> > Why would you want the inline `uid == 0` conditionals removed? They reflect the expected and correct behaviour for the tested code in that case. Namely, if somebody e.g. manually runs the unit tests as a privileged user.
> > I agree that we do **not** want to end up in that case running from our own `make` targets or unprivileged env runs, but I think that's a different matter. Similarly we should gradually phase out any remaining skip-decorators blindly using that condition now that we carefully handle privileges in `make test`.
> > If we want to actively refuse privileged use I think that would belong e.g. as an exception in the tested function and with tests adjusted accordingly.
> 
> Hm, ok. I had thought those checks were temporary to work around accidentally elevated privilege, my mistake if not.
> 
> However, I don't think having conditions like around the checked assertions is something we should really encourage. If there are behaviour differences for root users, and it's important to have those "documented", then I really think those ought to be separate tests asserting that circumsance/outcome. One could then much more cleanly make then conditional by using the standard skipping decorator, where you skip if uid ==0 in once case and uid !=0 in the other.

I had it the other way around in my head :upside_down_face: and would remove the `skip` decorators and keep the conditional. I don't really mind translating those tests with the uid-conditionals into two separate tests with corresponding "xor" `skip` decorators to always take either one or the other depending on `uid` if you prefer that. 
Yet, thinking it over again I have a nagging feeling that neither is great here. I mean, I believe we want to have unit tests closely verify existing behaviour of functions - even if suboptimal or not as such intended, as long as it is not directly buggy. Otherwise tests wont work as a canary in case of possible regressions. Still, I have a hunch that this behaviour was never really intended or desired but just not considered, because we generally run code as an unprivileged user and always have.
So I feel urged to introduce at least a log warning or even an exception in case those functions or perhaps even any central code is called as a privileged user. On the other hand I don't want this to turn into a completely different task with other potential regressions in case we actually have any use of it. E.g. if we actually ever end up running bits as root e.g. when checking which daemons to run in the init script or when launching sshd with paramiko hooks before yielding privileges.

Should we switch to `skip`s for now and add an issue + TODOs about looking into the underlying logic later or what do you think @albu-diku ?

albu-diku commented:
> > > > The tests look good - with the `uid == 0` conditionals removed throughout this can be approved and merged,
> > > 
> > > 
> > > Why would you want the inline `uid == 0` conditionals removed? They reflect the expected and correct behaviour for the tested code in that case. Namely, if somebody e.g. manually runs the unit tests as a privileged user.
> > > I agree that we do **not** want to end up in that case running from our own `make` targets or unprivileged env runs, but I think that's a different matter. Similarly we should gradually phase out any remaining skip-decorators blindly using that condition now that we carefully handle privileges in `make test`.
> > > If we want to actively refuse privileged use I think that would belong e.g. as an exception in the tested function and with tests adjusted accordingly.
> > 
> > 
> > Hm, ok. I had thought those checks were temporary to work around accidentally elevated privilege, my mistake if not.
> > However, I don't think having conditions like around the checked assertions is something we should really encourage. If there are behaviour differences for root users, and it's important to have those "documented", then I really think those ought to be separate tests asserting that circumsance/outcome. One could then much more cleanly make then conditional by using the standard skipping decorator, where you skip if uid ==0 in once case and uid !=0 in the other.
> 
> I had it the other way around in my head 🙃 and would remove the `skip` decorators and keep the conditional. I don't really mind translating those tests with the uid-conditionals into two separate tests with corresponding "xor" `skip` decorators to always take either one or the other depending on `uid` if you prefer that. Yet, thinking it over again I have a nagging feeling that neither is great here. I mean, I believe we want to have unit tests closely verify existing behaviour of functions - even if suboptimal or not as such intended, as long as it is not directly buggy. Otherwise tests wont work as a canary in case of possible regressions. Still, I have a hunch that this behaviour was never really intended or desired but just not considered, because we generally run code as an unprivileged user and always have. So I feel urged to introduce at least a log warning or even an exception in case those functions or perhaps even any central code is called as a privileged user. On the other hand I don't want this to turn into a completely different task with other potential regressions in case we actually have any use of it. E.g. if we actually ever end up running bits as root e.g. when checking which daemons to run in the init script or when launching sshd with paramiko hooks before yielding privileges.
> 
> Should we switch to `skip`s for now and add an issue + TODOs about looking into the underlying logic later or what do you think @albu-diku ?

To summarise where I landed in the group chat - as of the change to make sure we are running the tests as a standard user it will now be the _common_ case that is exercised. As such, while my sense is that ultimately "flattening" those tests may be worth it, it can be a follow-up relative to what is here and better that we get the checks running.

I will note though I think the disucssion was well worth it and this is rather an interesting case with respect to trade-offs and how to arrange things. I'm going to reflect further on what you described @jonasbardino as, at minimum, there might be clarifications and learnings to share etc.

albu-diku reviewed (approved):
Worth rebasing this atop latest next and getting one full test run, but am extending approval to the content of this PR as it is now.

jonasbardino commented:
> Worth rebasing this atop latest next and getting one full test run, but am extending approval to the content of this PR as it is now.

Thanks, I've rebased for testing again. I'll add the follow-up tag and create an issue about future work in the privileged user context.

jonasbardino commented:
@Martin-Rehr just reminded me that we _do_ need to run migrid code as a privileged user in a few cases like the new `grid_quota.py` daemon. So we may not be able to completely refuse fileio access checks as `uid == 0`, but we probably still want to warn about the pitfalls of any such use.
